src/kernel.ts

```diff
@@ -1,1069 +1,802 @@
-            cache: false,
-            type: "POST",
-            contentType: false,  // there's no data with this
-            dataType: "json",
-            success: this._on_success(on_success),
-            error: this._on_error(on_error)
-        });
-    }
+/*
+ * Kernel message specification.
+ */
+export
+interface IKernelMsg {
+  header: IKernelMsgHeader;
+  metadata: any;
+  content: any;
+  parentHeader: {} | IKernelMsgHeader;
+  msgId?: string;
+  msgType?: string;
+  channel?: string;
+  buffers?: string[] | ArrayBuffer[];
+}
 
-    reconnect(): void {
-        /**
-         * Reconnect to a disconnected kernel. This is not actually a
-         * standard HTTP request, but useful function nonetheless for
-         * reconnecting to the kernel if the connection is somehow lost.
-         *
-         * @function reconnect
-         */
-        if (this.is_connected()) {
-            return;
-        }
-        this._reconnect_attempt = this._reconnect_attempt + 1;
-        events.trigger('kernel_reconnecting.Kernel', {
-            kernel: this,
-            attempt: this._reconnect_attempt,
-        });
-        this.start_channels();
-    }
 
-    _on_success(success): any {
-        /**
-         * Handle a successful AJAX request by updating the kernel id and
-         * name from the response, and then optionally calling a provided
-         * callback.
-         *
-         * @function _on_success
-         * @param {function} success - callback
-         */
-        return (data, status, xhr) => {
-            if (data) {
-                this.id = data.id;
-                this.name = data.name;
-            }
-            this.kernel_url = utils.url_join_encode(this.kernel_service_url, this.id);
-            if (success) {
-                success(data, status, xhr);
-            }
-        };
-    }
+/**
+ * Object providing a Future interface for message callbacks.
+ *
+ * Only one callback can be registered per type.
+ * If `autoDispose` is set, the future will self-dispose after `isDone` is
+ * set and the registered `onDone` handler is called.
+ *
+ * The Future is considered done when a `reply` message and a
+ * an `idle` iopub status message have been received.
+ */
+export
+interface IKernelFuture {
+  /**
+   * Dispose and unregister the future.
+   */  
+  dispose(): void;
+
+  /**
+   * Register a reply handler. Returns `this`.
+   */
+  onReply(cb: (msg: IKernelMsg) => void): IKernelFuture;
+
+  /**
+   * Register an output handler. Returns `this`.
+   */
+  onOutput(cb: (msg: IKernelMsg) => void): IKernelFuture;
+
+  /**
+   * Register a done handler. Returns `this`.
+   */
+  onDone(cb: (msg: IKernelMsg) => void): IKernelFuture;
+
+  /**
+   * Register an input handler. Returns `this`.
+   */
+  onInput(cb: (msg: IKernelMsg) => void): IKernelFuture;
+
+  /**
+   * The autoDispose behavior of the future.
+   *
+   * If True, it will self-dispose() after onDone() is called.
+   */
+  autoDispose: boolean;
```

Comment:
I prefer public properties being defined above the public methods. Applies to `autoDispose` and `isDone`.